r/src/bpcells-cpp/matrixIterators/H5DenseMatrixWriter.h

```diff
@@ -0,0 +1,143 @@
+      }
+
+      if (file.exist(dataset_path)) {
+          file.unlink(dataset_path);
+      }
+      // Create the dataset
+      return file.createDataSet<T>(dataset_path, dataspace, props);
+    }
+
+  public:
+    H5DenseMatrixWriter(
+        const HighFive::File &file,
+        std::string path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t max_capacity = 16384,
+        uint64_t chunk_size = 1024,
+        uint32_t gzip_level = 0
+    )
+        : dataset(createH5Matrix(file, path, nrow, ncol, row_major, chunk_size, gzip_level))
+        , max_capacity(max_capacity) 
+        , row_major(row_major) {}
+
+    void write(MatrixLoader<T> &mat_in, std::atomic<bool> *user_interrupt = NULL) override {
+
+        // Ensure that we write matrices sorted by row
+        OrderRows<T> mat((std::unique_ptr<MatrixLoader<T>>(&mat_in)));
+        // Don't delete our original matrix
+        mat.preserve_input_loader();
+
+        uint32_t col = 0;
+        uint64_t idx = 0; // Index of for col_ptr array
+        std::vector<T> zero_buf(mat_in.rows());
+        std::vector<T> val_buf = zero_buf;  // buffer for each column
+        while (mat.nextCol()) {
+            if (user_interrupt != NULL && *user_interrupt) return;
+            if (mat.currentCol() < col)
+                throw std::runtime_error("H5DenseMatrixWriter encountered out-of-order columns");
+
+            while (mat.load()) {
+                uint32_t i = 0;
+                while (i < mat.capacity()) {
+                    uint32_t capacity = std::min((uint32_t)max_capacity, mat.capacity() - i);
+                    for (uint32_t ii = 0; ii < capacity; ii++) {
+                        val_buf[*(mat.rowData() + i + ii)] = *(mat.valData() + i + ii);
+                    }
+                    idx += capacity;
+                    i += capacity;
+                }
+                if (user_interrupt != NULL && *user_interrupt) return;
+            }
+            if (col == mat.currentCol() && idx > 0) {
+                if (row_major) {
+                    dataset.select({(uint64_t)mat.currentCol(), 0}, {1, val_buf.size()}).write_raw(val_buf.data(), datatype);
+                } else {
+                    dataset.select({0, (uint64_t)mat.currentCol()}, {val_buf.size(), 1}).write_raw(val_buf.data(), datatype);
+                }
+                col += 1;
+                val_buf = zero_buf;
+            }
+        }
+        dataset.createAttribute("encoding-type", std::string("array"));
+        dataset.createAttribute("encoding-version", std::string("0.2.0"));
+    }
+};
+
+// Write a Dense Array to an AnnData file
+template <typename T>
+H5DenseMatrixWriter<T> createAnnDataDenseMatrix(
+    std::string file,
+    std::string dataset,
+    uint32_t nrow,
+    uint32_t ncol,
+    bool row_major,
+    uint32_t buffer_size,
+    uint32_t chunk_size,
+    uint32_t gzip_level
+) {
```

Comment:
Could we re-arrange the code so most of the implementation can be in a `.cpp` file?

I think the best way to do this would be to only put the declaration for `createAnnDataDenseMatrix` in the header file, and make it return a `std::unique_ptr<MatrixWriter<T>>`, then move everything else into the `.cpp` file with explicit template instantiations, just like `createAnnDataMatrix` does ([link](https://github.com/bnprks/BPCells/blob/290863d148c3b9435a706c72ec6fb4e27db4bf07/r/src/bpcells-cpp/matrixIterators/ImportMatrixAnnDataHDF5.cpp#L498-L529)).

And as mentioned above, I think we'll probably want to get rid of the `nrow` and `ncol` parameters here and get them during the call to `write(mat_in)`